File: sms-frontend/types/grade.ts
// types/grade.ts

export interface GradeBase {
    name: string
}

export type GradeCreate = GradeBase

export interface GradeUpdate {
    name?: string
}

export interface GradeResponse extends GradeBase {
    id: string
    created_at?: string | null
    updated_at?: string | null
    created_by?: string | null
    updated_by?: string | null
}

export interface GradeDetailResponse extends GradeResponse {
    school?: Record<string, never> | null
    classes?: Record<string, never>[]
}

File: sms_back_end/database/multi_tenant_school_management/schemas/grade.py
from uuid import UUID
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, ConfigDict


# =========================
# BASE
# =========================

class GradeBase(BaseModel):
    name: str


# =========================
# CREATE
# =========================

class GradeCreate(GradeBase):
    pass


# =========================
# UPDATE
# =========================

class GradeUpdate(BaseModel):
    name: Optional[str] = None


# =========================
# SIMPLE CLASS (for relations)
# =========================

class ClassSimple(BaseModel):
    id: UUID
    name: str

    model_config = ConfigDict(from_attributes=True)


# =========================
# RESPONSE
# =========================

class GradeResponse(GradeBase):
    id: UUID

    # ✅ FIXED: use datetime instead of string
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    created_by: Optional[str] = None
    updated_by: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


# =========================
# DETAILED RESPONSE
# =========================

class GradeDetailResponse(GradeResponse):
    """
    Use when returning grade with relationships
    """

    # Avoid raw dicts — use typed models
    school: Optional[dict] = None

    # ✅ FIXED: no mutable default + proper typing
    classes: Optional[List[ClassSimple]] = None

    model_config = ConfigDict(from_attributes=True)